crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -128,64 +141,58 @@ pub struct ModuleResolutionSettings {
     /// The root of the workspace, used for finding first-party modules.
     pub workspace_root: FileSystemPathBuf,
 
+    /// Optional (already validated) path to standard-library typeshed stubs.
+    /// If this is not provided, we will fallback to our vendored typeshed stubs
+    /// bundled as a zip file in the binary
+    pub custom_typeshed: Option<FileSystemPathBuf>,
+
     /// The path to the user's `site-packages` directory, where third-party packages from ``PyPI`` are installed.
     pub site_packages: Option<FileSystemPathBuf>,
-
-    /// Optional path to standard-library typeshed stubs.
-    /// Currently this has to be a directory that exists on disk.
-    ///
-    /// (TODO: fall back to vendored stubs if no custom directory is provided.)
-    pub custom_typeshed: Option<FileSystemPathBuf>,
 }
 
 impl ModuleResolutionSettings {
     /// Implementation of PEP 561's module resolution order
     /// (with some small, deliberate, differences)
-    fn into_ordered_search_paths(self) -> OrderedSearchPaths {
+    fn into_ordered_search_paths(self) -> (SupportedPyVersion, OrderedSearchPaths) {
         let ModuleResolutionSettings {
+            target_version,
             extra_paths,
             workspace_root,
             site_packages,
             custom_typeshed,
         } = self;
 
-        let mut paths: Vec<_> = extra_paths
+        let mut paths = extra_paths
             .into_iter()
-            .map(|path| ModuleSearchPath::new(path, ModuleSearchPathKind::Extra))
-            .collect();
+            .map(ModuleResolutionPath::extra)
+            .collect::<Option<Vec<ModuleResolutionPath>>>()
+            .unwrap();
```

Comment:
Unwrapping isn't correct here; if the paths fail to validate, we need to bail out from the whole process and tell the user that we were passed an invalid path in the configuration they gave us. Will do that soon.

Reply:
I think this is where a builder for `ModuleSearchSettings` could help where `ModuleSearchSettings` stores `ModuleResolutionPath` internally but the builder exposes `with_extra(extras) -> Result<Self, Self>` methods.

Reply:
Possibly, but I'm still not sure I like that. I'd prefer to defer that to a followup, anyway. I'll add a TODO here.

Reply:
Nit: 

```
let mut paths: Vec<_> = extra_paths
            .into_iter()
            .map(|path| ModuleResolutionPathBuf::extra(path).unwrap())
            .collect();
```

Makes it more obvious which path failed to unwrap.